proposals/open-ended-ranges.md

```diff
@@ -0,0 +1,260 @@
+allowing to instantiate a range of discrete integral values like `0 until size` which is equivalent to `0..(size - 1)`.
+
+Despite all of this, due to asymmetry between `..` and `until`, the former is used more often, even in cases where the latter would be more clear.
+
+We have conducted a UX research that showed that Kotlin users may experience troubles identifying whether the ranges created with 
+the `..` operator and `until` function have their upper bound excluded or included. We also evaluated how introducing new operators
+for range expressions may improve this situation, and it turned out that the effect of that can be significant: 
+the misinterpretation rate was less by ~20-30% with them.
+
+For the above reason, we propose to introduce the operator `..<` in Kotlin that would be 
+on par with the `..` operator and make it very clear that the upper bound is not included.
+
+## Use cases
+
+Currently, the use cases of the new operator are mostly covered by the `until` function, that corrects the upper bound
+returning a closed range of integral values that would be equivalent to an open-ended range. 
+However, the `until` function is available only for integral or discrete argument types, such as `Int`, `Long`, `Char`, 
+and having the new operator gives a chance for introducing open-ended ranges for those type that didn't have it before.
+
+### Iterating indices of a data structure
+
+Typically, data structures start indexing at zero and thus most indexed loops on these data structures
+have the form of `for (index in 0 until size)`. Such a common use case deserves introducing a designated operator with
+a clear meaning.
+
+### Discretization and bucketing of continuous values
+
+Discretization and binning require splitting the domain of a continuous value to a number of non-overlapping intervals.
+Such intervals are usually chosen as ranges that include their lower bound and exclude the upper bound, so that two 
+adjacent ranges neither have a point where they overlap, nor a point between them that is not contained in these ranges.  
+
+Even sometimes when the value is already discrete, for example, when it is expressed as a `Double` number, 
+and it is possible to represent a half-open range with a closed one by adjusting one of its bounds, 
+in practice it is not convenient to work with such ranges:
+
+```kotlin
+val equivalent = 1.0..2.0.nextDown() // contains the same values as 1.0..<2.0 range
+println(eqivalent)  // 1..1.9999999999999998
+```
+
+## Similar API review
+
+### Languages that distinguish end-inclusive and end-exclusive ranges
+
+- Swift: `...` end-inclusive range, `..<` end-exclusive, supports one-sided ranges 
+- Ruby: `..` end-inclusive range, `...` end-exclusive, supports one-sided ranges
+- Groovy: `..` end-inclusive range, `..<` end-exclusive, `<..` start-exclusive, `<..<` both bounds-exclusive range
```

Comment:
A Rust is also worth noting here to add a variety to the syntax that different languages use (Rust is an outlier in using `..` for an end-exclusive range).